File: modules/module-provider/src/main/scala/io/constellation/provider/sdk/ModuleExecutorServer.scala
package io.constellation.provider.sdk

import cats.effect.{IO, Ref}

import io.constellation.CValue
import io.constellation.provider.CValueSerializer
import io.constellation.provider.v1.{provider => pb}

/** Dispatches incoming ExecuteRequest to the correct ModuleDefinition handler.
  *
  * Module list is stored in a Ref to support hot-swap (canary rollout).
  */
class ModuleExecutorServer(
    modules: Ref[IO, List[ModuleDefinition]],
    serializer: CValueSerializer
) {

  /** Handle an ExecuteRequest by dispatching to the appropriate module. */
  def handleRequest(request: pb.ExecuteRequest): IO[pb.ExecuteResponse] =
    for {
      startTime <- IO.monotonic
      response  <- dispatch(request).handleError {
        case e: ModuleNotFoundException => e.errorResponse
        case e: TypeErrorException      => e.errorResponse
        case error =>
          pb.ExecuteResponse(
            result = pb.ExecuteResponse.Result.Error(pb.ExecutionError(
              code = "RUNTIME_ERROR",
              message = error.getMessage,
              stackTrace = error.getStackTrace.take(10).mkString("\n")
            ))
          )
      }
      endTime <- IO.monotonic
      durationMs = (endTime - startTime).toMillis
      withMetrics = response.copy(
        metrics = Some(pb.ExecutionMetrics(durationMs = durationMs))
      )
    } yield withMetrics

  private def dispatch(request: pb.ExecuteRequest): IO[pb.ExecuteResponse] =
    for {
      currentModules <- modules.get
      moduleDef <- currentModules.find(_.name == request.moduleName) match {
        case Some(md) => IO.pure(md)
        case None     => IO.raiseError(new ModuleNotFoundException(request.moduleName))
      }
      inputCValue <- IO.fromEither(
        serializer.deserialize(request.inputData.toByteArray).left.map(e =>
          new TypeErrorException(s"Failed to deserialize input: $e")
        )
      )
      outputCValue <- moduleDef.handler(inputCValue)
      outputBytes <- IO.fromEither(
        serializer.serialize(outputCValue).left.map(e =>
          new RuntimeException(s"Failed to serialize output: $e")
        )
      )
    } yield pb.ExecuteResponse(
      result = pb.ExecuteResponse.Result.OutputData(
        com.google.protobuf.ByteString.copyFrom(outputBytes)
      )
    )

  /** Convert this server into a request handler function for use with ExecutorServerFactory. */
  def toHandler: pb.ExecuteRequest => IO[pb.ExecuteResponse] = handleRequest
}

private[sdk] class ModuleNotFoundException(moduleName: String)
    extends RuntimeException(s"Module not found: $moduleName") {
  def errorResponse: pb.ExecuteResponse = pb.ExecuteResponse(
    result = pb.ExecuteResponse.Result.Error(pb.ExecutionError(
      code = "MODULE_NOT_FOUND",
      message = s"Module not found: $moduleName"
    ))
  )
}

private[sdk] class TypeErrorException(msg: String) extends RuntimeException(msg) {
  def errorResponse: pb.ExecuteResponse = pb.ExecuteResponse(
    result = pb.ExecuteResponse.Result.Error(pb.ExecutionError(
      code = "TYPE_ERROR",
      message = msg
    ))
  )
}

File: sdks/typescript/src/provider/module-executor-server.ts
/**
 * Dispatches incoming ExecuteRequests to the correct ModuleDefinition handler.
 *
 * Mirrors the Scala ModuleExecutorServer class.
 */

import type { ModuleDefinition } from "../types/module-definition.js";
import type { CValueSerializer } from "../serialization/cvalue-serializer.js";
import type { ExecuteRequest, ExecuteResponse } from "../transport/transport.js";

export class ModuleExecutorServer {
  private modules: ModuleDefinition[];
  private readonly serializer: CValueSerializer;

  constructor(modules: ModuleDefinition[], serializer: CValueSerializer) {
    this.modules = [...modules];
    this.serializer = serializer;
  }

  /** Update the module list (for canary rollout). */
  setModules(modules: ModuleDefinition[]): void {
    this.modules = [...modules];
  }

  /** Handle an ExecuteRequest by dispatching to the appropriate module. */
  async handleRequest(request: ExecuteRequest): Promise<ExecuteResponse> {
    const startTime = performance.now();
    try {
      const response = await this.dispatch(request);
      const durationMs = Math.round(performance.now() - startTime);
      return {
        ...response,
        metrics: { durationMs, memoryBytes: 0 },
      };
    } catch (error) {
      const durationMs = Math.round(performance.now() - startTime);
      const err = error instanceof Error ? error : new Error(String(error));
      return {
        result: {
          case: "error",
          value: {
            code: "RUNTIME_ERROR",
            message: err.message,
            stackTrace: err.stack?.split("\n").slice(0, 10).join("\n") ?? "",
          },
        },
        metrics: { durationMs, memoryBytes: 0 },
      };
    }
  }

  /** Convert this server into a request handler function for use with ExecutorServerFactory. */
  toHandler(): (request: ExecuteRequest) => Promise<ExecuteResponse> {
    return (request) => this.handleRequest(request);
  }

  private async dispatch(request: ExecuteRequest): Promise<ExecuteResponse> {
    const moduleDef = this.modules.find((m) => m.name === request.moduleName);
    if (!moduleDef) {
      return {
        result: {
          case: "error",
          value: {
            code: "MODULE_NOT_FOUND",
            message: `Module not found: ${request.moduleName}`,
            stackTrace: "",
          },
        },
      };
    }

    let inputCValue;
    try {
      inputCValue = this.serializer.deserialize(request.inputData);
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      return {
        result: {
          case: "error",
          value: {
            code: "TYPE_ERROR",
            message: `Failed to deserialize input: ${msg}`,
            stackTrace: "",
          },
        },
      };
    }

    const outputCValue = await moduleDef.handler(inputCValue);

    let outputBytes;
    try {
      outputBytes = this.serializer.serialize(outputCValue);
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      throw new Error(`Failed to serialize output: ${msg}`);
    }

    return {
      result: { case: "outputData", value: outputBytes },
    };
  }
}
